### ISSUE TITLE
What states can tasks be in?

### ISSUE THREAD

**Issue by nh13** [fulcrumgenomics/dagr#123]
I would like to start a discussion about what states a task can and should be in relative to their execution in dagr.
### Current Implementation

For posterity, I will first describe the current implementation.  There are two places status information is stored about the execution of a task: `TaskStatus` and `GraphNodeState`.  `TaskStatus` has the following states:
1. `UNKNOWN`: The task state is unknown (in reality, this is the inital state of the task until it is in `STARTED`).
2. `STARTED`: The task has been started (`TaskExecutionRunner` is running the task).
3. `STOPPED`: The task has been stopped (`TaskExecutionRunner` terminated it).
4. `FAILED_GET_TASKS`: The task could not get the list of tasks (`Task.getTasks` threw an `Exception`).
5. `FAILED_SCHEDULING`: The task could not execute after scheduling (any `Exception` was thrown while trying to start the task in `TaskExecutionRunner.runTask`).
6. `FAILED_COMMAND`: The task command has failed (the task returned a non-zero exit code in `TaskRunnable`)
7. `FAILED_ON_COMPLETE`: The task has failed when trying to run its onComplete method (see `onCompleteSuccessful` in `TaskRunnable`).
8. `SUCCEEDED`: The task has succeeded.
9. `MANUALLY_SUCCEEDED`: The task was manually succeeded.
10. 

For non-`UnitTask`s, it doesn't enter in 4-7 since it simply succeeds when all its children succeed.  Now for `GraphNodeState` we have:
1. `ORPHAN`: has predecessors that have not been seen by the system (this means that it depends on a task that isn't tracked by the task manager).
2. `PREDECESSORS_AND_UNEXPANDED`: has predecessors (it has unmet dependencies) and [[Task.getTasks()]] has not been called (it has unmet dependencies)
3. `ONLY_PREDECESSORS`: has predecessors (it has unmet dependencies) after [[dagr.core.tasksystem.Task.getTasks()]] was called.
4. `NO_PREDECESSORS`: no predecessors, eligible for scheduling (execution).
5. `RUNNING`: is running.
6. `COMPLETED`: is completed (it could have succeeded or failed, but it means don't try to schedule or execute).
### The Future

It would be nice to get merge `TaskStatus` and `GraphNodeState` for the maintenance of the execution graph.  We may still want to have some internal status that isn't exposed ot the user, but I am not sure what that would be just yet, besides not showing all the failure states (which I would not advocate doing).

I think it would be also nice to find a different way to update the status of `Pipelines` when their children complete.  Currently, we build the pipeline and make the pipeline dependent on any child task, and put the pipeline into the `ONLY_PREDECESSORS` state.

I think user-facing statuses could be:
1. `UNKNOWN_DEPENDENCIES` - a task that has depedencies that have not been added to the task manager (previously `ORPHAN).
2. `UNMET_DEPENDENCIES` - a task that has dependencies that have not completed.
3. `ELIGIBLE` - a task that has no unmet dependencies and eligible for scheduling.
4. `RUNNABLE` - a task that has been scheduled for execution but has not begun execution.
5. `RUNNING` - a task that is currently being executed.
6. `FAILED` - a task that failed (for any reason).
7. `SUCCEEDED` - a task that executed successfully.

I think that we could have more granular failure states as we do now, that tell folks _why_ the task failed, or we could just keep a status message.  I like keeping an internal state for failure since we may choose to leverage that going forward, for example retrying the `onComplete` method instead of executing the task again.

We could also store states for when the user stops a task (ex. `STOPPED`), moves a task from `FAILED` to `SUCCEEDED` (ex. `MANUALLY_SUCCEEDED`), or a state for failure that allows for dependencies to still run.

**Comment by nh13** [comment]
@tfenne I think it would be nice to review the `TaskManager` in person to talk about how the execution works, and then start working on this discussion.  In reality, I think if we choose to re-implement the execution of tasks using an actor model, the state diagram becomes really important to design first.

**Comment by tfenne** [comment]
@nh13 I agree with most of this, and am happy to talk in person.  A few opening thoughts:
1. Yay for merging `TaskStatus` and `GraphNodeStatus`, since I'm continually confused by the separation.
2. I think we should re-implement the statuses as a sealed trait and case classes instead of an enum.  This way we can add, e.g. subclasses of `Failed` and `Succeeded`, and code that cares about the specific sub-type can care, but places that don't can just match on `Failed` and `Succeeded` as catch-alls for those types.  Also, as case classes, they can hold messages and further information as we see fit, which an Enumeration can't.
3. I'm fairly sure you'll disagree with this, but I think the task manager should assert that as soon as scheduling/execution begins that all task's predecessors are registered, making `UNKNOWN_DEPENDENCIES` unnecessary.
4. I do think the sub-types of failure are useful in, e.g. the execution report, as it helps narrow down where things went wrong.
5. Is there a logical gap such that `Runnable` is really necessary?  I think that as the task is handed off to the `TaskRunner` it could just go into `Running`.  Or to play the other side, should there be a 'Completable' status between `Running` and `Failed|Succeeded`?

**Comment by nh13** [comment]
RE (3): I am actually using this in production so lets be careful to find a way around what I am using it for.  It's related to wanting Pipelines to have their `build` method called prior to any of those Pipelines executing.

RE (5): I only think `Runnable` is necessary if when we re-implement the execution system such that it is possible to have a failure between `Eligible` and `Running`.  Right now it's not an issue.
